src/sciline/container.py

```diff
@@ -14,81 +25,128 @@ class UnsatisfiedRequirement(Exception):
     pass
 
 
-class Container:
-    def __init__(self, inj: injector.Injector, /, *, lazy: bool) -> None:
-        self._injector = inj
-        self._lazy = lazy
-
-    def get(self, tp: Type[T], /) -> Union[T, Delayed]:
-        try:
-            # We are slightly abusing Python's type system here, by using the
-            # injector to get T, but actually it returns a Delayed that can
-            # compute T. self._injector does not know this due to how we setup the
-            # bindings. We'd like to use Delayed[T], but that is not supported yet:
-            # https://github.com/dask/dask/pull/9256
-            task: Delayed = self._injector.get(tp)  # type: ignore
-        except injector.UnsatisfiedRequirement as e:
-            raise UnsatisfiedRequirement(e) from e
-        return task if self._lazy else task.compute()
-
-
-def _delayed(func: Callable) -> Callable:
+class AmbiguousProvider(Exception):
+    pass
+
+
+def _is_compatible_type_tuple(
+    requested: tuple[type, ...], provided: tuple[type | TypeVar, ...]
+) -> bool:
     """
-    Decorator to make a function return a delayed object.
+    Check if a tuple of requested types is compatible with a tuple of provided types.
 
-    In contrast to dask.delayed, this uses functools.wraps, to preserve the
-    type hints, which is a prerequisite for injector to work.
+    Types in the tuples must either by equal, or the provided type must be a TypeVar.
     """
-    import dask
+    for req, prov in zip(requested, provided):
+        if isinstance(prov, TypeVar):
+            continue
+        if req != prov:
+            return False
+    return True
 
-    @wraps(func)
-    def wrapper(*args, **kwargs):
-        return dask.delayed(func)(*args, **kwargs)
 
-    return wrapper
+def _bind_free_typevars(tp: type, bound: Dict[TypeVar, type]) -> type:
+    if isinstance(tp, TypeVar):
+        return bound[tp]
```

Comment:
Shouldn't it also be bound.get(tp, tp) in case it's not bound to anything, or is that not possible at all?

Reply:
I want it to fail in this case, so `__getitem__` works (to get `KeyError`). I could use what you suggest, to get an `UnsatisfiedRequirement` elsewhere. Not sure which is better? Or maybe we should have another specific exception like `UnboundTypeVar`?

Reply:
I think raising `UnboundTypeVar` makes more sense because that's what you'll get from static type checker also.